src/query/sortResults.ts

```diff
@@ -0,0 +1,91 @@
+import debug from 'debug'
+import get from 'lodash.get'
+import { InternalEntity } from 'src/glossary'
+import { OrderBy, SortDirection } from './queryTypes'
+
+const log = debug('sortResults')
+
+type FlatSortCriteria = [string[], SortDirection]
+
+export function warnOnIneffectiveSortingKeys(obj: Record<string, any>): void {
+  const [mainCriteria, ...siblings] = Object.keys(obj)
+
+  if (siblings.length > 0) {
+    console.warn(
+      'Sorting by "%s" has no effect: already sorted by "%s".',
+      siblings.join(','),
+      mainCriteria,
+    )
+  }
+}
+
+function flattenSortCriteria(
+  orderBy: OrderBy<any>[],
+  propertyPath: string[] = [],
+): FlatSortCriteria[] {
+  log('flattenSortCriteria:', orderBy, propertyPath)
+
+  return orderBy.reduce<FlatSortCriteria[]>((criteria, properties) => {
+    warnOnIneffectiveSortingKeys(properties)
+
+    // Multiple properties in a single criteria object are forbidden.
+    // Use the list of criteria objects for multi-criteria sort.
+    const property = Object.keys(properties)[0]
+    const sortDirection = properties[property]!
+    const path = propertyPath.concat(property)
+    log({ property, sortDirection, path })
+
+    // Recursively flatten order criteria when referencing
+    // relational properties.
+    const newCriteria =
+      typeof sortDirection === 'object'
+        ? flattenSortCriteria([sortDirection], path)
+        : [[path, sortDirection]]
+
+    log('pushing new criteria:', newCriteria)
+    return criteria.concat(newCriteria as any)
+  }, [])
+}
+
+export function sortResults(
+  orderBy: OrderBy<any> | OrderBy<any>[],
+  data: InternalEntity<any, any>[],
+): void {
+  log('sorting data:', data)
+  log('order by:', orderBy)
+
+  const criteriaList = ([] as OrderBy<any>[]).concat(orderBy)
+  log('criteria list:', criteriaList)
+
+  const criteria = flattenSortCriteria(criteriaList)
+  log('flattened criteria:', JSON.stringify(criteria))
+
+  data.sort((left, right) => {
+    for (const [path, sortDirection] of criteria) {
+      log('sorting "%s" ("%s")', path, sortDirection)
```

Comment:
this log is probably not needed given the log on line 70 (unless lodash.get has a really terrible error message if it explodes)

Reply:
also, printing `n * log(n) * numberOfCriteria` messages can be a bit too much logging even for `debug` level 😬

sorting tens of thousands of records without logs should be still fast,, but trying to log it might make it super slow - do you expect `debug` level to be used only by library maintainers (=> ok to debug in the loop) or is it a feature for users (=> might be better to hide the logging behind an extra flag)?